### PR TITLE
remove functions from hard dependency

### PR DESCRIPTION
**Pull request by Sneezry**

### CODE REVIEW

File: src/Models/AzureFunctions.ts [hunk 1]
```diff
@@ -185,6 +208,11 @@ export class AzureFunctions implements Component, Provisionable, Deployable {
   }
 
   async provision(): Promise<boolean> {
+    const isFunctionsExtensionAvailable = await AzureFunctions.isAvailable();
```
Comment (lirenhe) [inline src/Models/AzureFunctions.ts]: Is it possible to check whether the project contains azure function and then ask user to install it?

Comment (Sneezry) [reply src/Models/AzureFunctions.ts]: Yeah. `AzureFunctions.isAvailable` is a static method, so we can call it anywhere without context.

I thinks this is more like a user experience topic. Should we show the notification as soon as the user opens a project contains Functions component or only when user executes provision command?

File: src/Models/AzureFunctions.ts [hunk 2]
```diff
@@ -85,6 +85,33 @@ export class AzureFunctions implements Component, Provisionable, Deployable {
     return this.componentType;
   }
 
+  static async isAvailable(): Promise<boolean> {
+    if (!vscode.extensions.getExtension(
+            'ms-azuretools.vscode-azurefunctions')) {
```
Comment (lirenhe) [inline src/Models/AzureFunctions.ts]: Move it to constants

File: src/Models/AzureFunctions.ts [hunk 3]
```diff
@@ -85,6 +85,33 @@ export class AzureFunctions implements Component, Provisionable, Deployable {
     return this.componentType;
   }
 
+  static async isAvailable(): Promise<boolean> {
+    if (!vscode.extensions.getExtension(
+            'ms-azuretools.vscode-azurefunctions')) {
+      const choice = await vscode.window.showInformationMessage(
+          'Azure Functions extension has been disabled or not installed. Please enable Azure Functions extension or install it from marketplace, and try again.',
```
Comment (lirenhe) [inline src/Models/AzureFunctions.ts]: Is it better to use:    Azure Functions extension is required for the current project. Do you want to install it from marketplace?

File: src/Models/AzureFunctions.ts [hunk 4]
```diff
@@ -85,6 +85,33 @@ export class AzureFunctions implements Component, Provisionable, Deployable {
...
+    if (!vscode.extensions.getExtension(
+            'ms-azuretools.vscode-azurefunctions')) {
+      const choice = await vscode.window.showInformationMessage(
+          'Azure Functions extension has been disabled or not installed. Please enable Azure Functions extension or install it from marketplace, and try again.',
+          'Install Azure Functions Extension', 'Close');
+      if (choice === 'Install Azure Functions Extension') {
+        vscode.commands.executeCommand(
+            'vscode.open',
+            vscode.Uri.parse(
+                'vscode:extension/ms-azuretools.vscode-azurefunctions'));
```
Comment (lirenhe) [inline src/Models/AzureFunctions.ts]: ms-azuretools.vscode-azurefunctions should be a constant value

File: src/Models/IoTProject.ts [hunk 5]
```diff
@@ -143,6 +143,10 @@ export class IoTProject {
         this.componentList.push(functionApp);
       }
 
+      this.componentList.forEach(item => {
+        item.checkPrerequisites();
```
Comment (lirenhe) [inline src/Models/IoTProject.ts]: checkPrerequisites is async function

Comment (Sneezry) [reply src/Models/IoTProject.ts]: The order is not necessary here, so `await` is not required.

**Pull request merged**